rts/motoko-rts/src/gc/incremental/roots.rs

```diff
@@ -1,75 +1,65 @@
-use crate::{types::Value, visitor::pointer_to_dynamic_heap};
+use motoko_rts_macros::ic_mem_fn;
+
+use crate::{types::Value, visitor::is_non_null_pointer_field};
+
+/// An array referring to the static program variables, being
+/// - All canister variables.
+/// - Pooled shared objects.
+/// The array constitutes a GC root that is reinitialized on each canister upgrade.
+/// The scalar sentinel denotes an uninitialized root.
+#[cfg(feature = "ic")]
+static mut STATIC_VARIABLES: Value = Value::from_scalar(0);
 
 /// GC root set.
-#[derive(Clone, Copy)]
-pub struct Roots {
-    pub static_roots: Value,
-    pub continuation_table_location: *mut Value,
-    pub region0_ptr_location: *mut Value,
-    // If new roots are added in future, extend `visit_roots()`.
-}
+pub type Roots = [*mut Value; 6];
 
 #[cfg(feature = "ic")]
 pub unsafe fn root_set() -> Roots {
-    use crate::memory::ic;
-    Roots {
-        static_roots: ic::get_static_roots(),
-        continuation_table_location: crate::continuation_table::continuation_table_loc(),
-        region0_ptr_location: crate::region::region0_get_ptr_loc(),
-    }
+    use crate::{
+        continuation_table::continuation_table_loc,
+        persistence::{stable_actor_location, stable_type_descriptor},
+        region::region0_get_ptr_loc,
+    };
+    [
```

Comment:
Will rust detect a mismatch between the size of this array and the type of `Roots` above? If not, maybe define a const ROOTS_SIZE and an assert using it.

Reply:
Yes, the Rust compiler checks the size of the array expression to match the size of the array type.